r1cs/gadgets/std/src/fields/nonnative/reduce.rs

```diff
@@ -252,25 +424,95 @@ impl<SimulationF: PrimeField, ConstraintF: PrimeField> Reducer<SimulationF, Cons
             ));
         }
 
-        // This part we mostly use the techniques in bellman-bignat
-        // The following code is adapted from https://github.com/alex-ozdemir/bellman-bignat/blob/master/src/mp/bignat.rs#L567
+        // The equality of two `A`-ary representations `[L[0],L[1],..]` and `[R[0],R[1],..]` with 
+        // oversized limbs is proven by enforcing their *shifted* differences 
+        // ``
+        //      [L[0] + (- R[0] + shift_constant) , 
+        //          L[1] + (- R[1] + shift_constant), 
+        //              ..],
+        // `` 
+        // to represent 
+        // ``
+        //      Sum_{i>=0} shift_constant * A^i.
+        // ``
+        // The constant `shift_constant = 2^bits_per_group  - 1` is to circumvent underflows in a
+        // length-preserving manner. With this choice  
+        // ``
+        //     0 <=  shift_constant - R[i] < 2^bits_per_group,
+        // `` 
+        // and hence
+        // ``
+        //      L[i] + (shift_constant - R[i]) < 2^{bits_per_group + 1},
+        // ``
+        // Since the length of the carries are throughout `<= bits_per_limb + surfeit`, 
+        // we also have `carry[i-1] + L[i] < 2^bits_per_group`, and the overall sum 
+        // ``
+        //      carry[i-1] + L[i] + (shift_constant - R[i]) 
+        // `` 
+        // is still of length `<= bits_per_group + 1`.
+        
+        // Why carries don't effect the length bound: 
+        // Assuming `len(carry) <= bits_per_limb + surfeit`, which is shown in detail below, 
+        // `` 
+        //   carry + limb[0] + limb[1] * A < 2 * A * 2^limb_size
+        // ``
+        // and therefore 
+        // ``
+        //   carry + limb[0] + limb[1] * A + limb[2] * A^2 < 2 * A^2 * 2^limb_size  
+        //   ...
+        //   carry + limb[0] + limb[1] * A + ... + limb[S-1] * A^{S-1} <
+        //                       < 2 * A^{S-1} * 2^limb_size.
+        // ``
+        // Thus
+        // ``
+        //   len(carry + L[i]) <= 
+        //              limb_size + 1 + (S-1) * len(A) = bits_per_group.
+        // `` 
+
+        // The following code is ported from [[Bellman]].
+        // [Bellman]: https://github.com/alex-ozdemir/bellman-bignat/blob/master/src/mp/bignat.rs#L567
         let mut carry_in = zero;
         let mut carry_in_value = ConstraintF::zero();
         let mut accumulated_extra = BigUint::zero();
+        // The group totals have arity `A^S = 2^{S*shift_per_limb}`.
         for (group_id, (left_total_limb, right_total_limb, num_limb_in_this_group)) in
             groupped_limb_pairs.iter().enumerate()
         {
+            // Carry and remainder are subject to the following constraints: 
+            // The quotient-remainder constraint
+            // ``
+            //    carry_in + group_total_left + shift_constant - group_total_right 
+            //          == carry * A^S + 0 + (shift_constant % A^S),
+            // ``
```

Comment:
The constraint should be: `carry_in + group_total_left + shift_constant - group_total_right == carry* A^S + 0 +(accumulated_extra + shift_constant % A^S)

Reply:
I agree. This was not well explained in the comments. I will add the explanation.

Reply:
Done in in PR #164.